reference_submissions/librispeech_deepspeech/librispeech_jax/submission.py

```diff
@@ -0,0 +1,187 @@
+  return 256
+
+
+def optimizer(hyperparameters: spec.Hyperparameters, num_train_examples: int):
+  opt_init_fn, opt_update_fn = optax.inject_hyperparams(optax.adamw)(
+      b1=hyperparameters.beta1,
+      b2=hyperparameters.beta2,
+      eps=hyperparameters.epsilon,
+      weight_decay=hyperparameters.weight_decay,
+      learning_rate=0.0)
+  return opt_init_fn, opt_update_fn
+
+
+def init_optimizer_state(workload: spec.Workload,
+                         model_params: spec.ParameterContainer,
+                         model_state: spec.ModelAuxiliaryState,
+                         hyperparameters: spec.Hyperparameters,
+                         rng: spec.RandomState) -> spec.OptimizerState:
+  del model_state
+  del rng
+
+  params_zeros_like = jax.tree_map(lambda s: jnp.zeros(s.shape_tuple),
+                                   workload.param_shapes)
+  opt_init_fn, opt_update_fn = optimizer(hyperparameters,
+                                         workload.num_train_examples)
+  optimizer_state = opt_init_fn(params_zeros_like)
+  return jax_utils.replicate(optimizer_state), opt_update_fn
+
+
+def l2_regularization(params, l2_decay_rank_threshold):
+  """Computes the squared l2 norm of the given parameters.
+
+  This function will only filter for parameters with
+  rank >= l2_decay_rank_threshold. So if this threshold is set to 2, then all
+  1d (and lower) parameter arrays, including all bias and batch norm params,
+  will be ignored in this computation.
+
+
+  Args:
+    params: Pytree containing parameters.
+    l2_decay_rank_threshold: The calculation will only include parameters with
+       param.ndim >= l2_decay_rank_threshold. Set to 2 to ignore all bias and
+       batch_norm params in the model.
+
+  Returns:
+    weight_l2: the squared l2 norm of all params matching the threshold.
+  """
+  weight_penalty_params = jax.tree_leaves(params)
+  weight_l2 = sum(
+      jnp.sum(x**2)
+      for x in weight_penalty_params
+      if x.ndim >= l2_decay_rank_threshold)
+  return weight_l2
+
+
+def update_step(batch,
+                params,
+                batch_stats,
+                optimizer_state,
+                workload,
+                global_step,
+                hyperparameters,
+                opt_update_fn,
+                rng,
+                lr):
+
+  optimizer_state.hyperparams['learning_rate'] = lr
+
+  def _loss_fn(params):
+    """loss function used for training."""
+    params_rng, dropout_rng = jax.random.split(rng, 2)
+    (logits, logit_paddings), new_batch_stats = workload.model_fn(
+        params,
+        batch,
+        batch_stats,
+        spec.ForwardPassMode.TRAIN,
+        {'params' : params_rng, 'dropout' : dropout_rng})
+
+    loss = workload.loss_fn(batch['targets'], (logits, logit_paddings))
+    return loss, new_batch_stats
+
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  (loss, new_model_state), grad = grad_fn(params)
+
+  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```

Comment:
(we can follow up offline too, but)

is it possible to flip the order of these lines?

```
  loss, grad = lax.pmean((loss, grad), axis_name='batch')
  grad_norm = jnp.sqrt(l2_regularization(grad, 0))
```

Reply:
fixed it in latest deepspeech submission code, now pmean happens before grad clipping